### DISCUSSION TITLE
RFC-4: persistent worker

### DISCUSSION THREAD

**Discussion by eed3si9n** [sbt/sbt#7653] [Ideas]
https://eed3si9n.com/rfc-4-persistent-worker/

**Comment by adpi2** [comment]
Thanks Eugene for the write up.

While I agree with the general idea of running jobs in the background, I am bit concerned about forking workers as processes. And I don't think that should be a goal for sbt 2.0.

RFC-4 intends to fix two issues:

1. BSP responsiveness
2. JIT vs stability

I think we should fix those two issues separately, in two steps: we ship a fix for (1) first and we fix (2) later, ideally after sbt 2.0 release.

To fix (1) we don't need forking. What we need is a better/simpler `BackgroundJobService` to implement `run`, `test`, `console` and `compile`:

- an intuitive interface for the user
- automatic isolation of input files (e.g. classpath)
- forwarded inputs/outputs to the client
- cancellation
- etc

Also I think we can rethink the UX of background jobs, in the light of client/server architecture:

- A background job should block the client if it's a shell. To launch another job the user can start another shell client.
- Do we still need the `bgList`, `bgStop` tasks? Yes, maybe to cancel jobs started by a BSP client.
- Probably we can remove `bgRun`, `bgTest`. It would be the same as running `run` or `test` in a separate client.

Knowing little about the current `BackgroundJobService`, I don't know how far we are from the goal. Maybe not so far, which would be a good news. But still I won't start working on (2), before (1) is shipped. Otherwise it's going to be a lot more work before we can release, a lot more unstability, and bugs to fix. The more processes, the harder it is to observe the program and to debug it when something goes wrong.

Two more things:

- Forked workers can only support a set of predefined jobs: the protocol. And so a user cannot create its own custom job. So even if we have persistent workers, we still need a in-process `BackgroundJobService` for custom jobs. Some candidates for this are Scala.js linker, assembling, shading...
- A persistent worker for `run`, `test`, `console`, `compile`: it looks like Bloop right? But the problem of Bloop is that we need to export the entire build configuration first. It would be simpler to send self-defined task definition as you described in the RFC-4. If we had such a Bloop, it would make Scala CLI simpler too.

**Reply by eed3si9n** [reply]
> - Forked workers can only support a set of predefined jobs: the protocol. And so a user cannot create its own custom job. 

I've added a section called 'persistent worker as a plugin task isolation' - https://eed3si9n.com/rfc-4-persistent-worker/. Basically plugin authors can implement a CLI in any Scala version, the plugin can run the CLI to perform the task. One neat side effect of this approach is complete classpath isolation so plugin can do whatever, and metabuild won't get affected.

**Reply by adpi2** [reply]
I like this approach. It looks like the Coursier app format.

But still, it seems as a first step we can run everything in a single process, for the sake of observability.

**Comment by eed3si9n** [comment]
Replying to @BillyAutrey's comment on https://github.com/sbt/sbt/issues/7643#issuecomment-2353433832,

> The RFP talks about the API that a persistent worker would use, which is good. The implementation of a basic worker makes sense, and it wouldn't be too bad. But, the desired UX/outcome is still a question here. 

First, regarding stability,  what we have observed is that [forking to an external JVM process during testing](https://www.scala-sbt.org/1.x/docs/Forking.html) as

```scala
Test / fork := true
```

makes tests more stable. On GitHub there are [9.4k](https://github.com/search?q=%22fork+%3A%3D+true%22+path%3A*.sbt&type=code&p=1) builds with some kind of `fork := true`. We don't enable this by default partly because of slow startup, and probably because forking handles one subproject at a time.

https://eed3si9n.com/rfc-4-persistent-worker/ lists two problems:

- **BSP responsiveness issue**: Since the idea of sbt server has been proposed to handle requests from both human users and IDEs, we’ve known that blocking on all commands were inconvenient. While `test` runs, sbt cannot respond to Metals, which results in Metals shutting down sbt.
- **JIT vs stability**: This is a good segue into the process emulation that sbt 1.x performs. Instead of shelling out a new JVM each time, sbt 1.x creates a sandbox classloader for test and run and runs in the same JVM as the sbt process. This has the benefit of caching the HotSpot JITing. On the other hand, letting test run in the same JVM process causes various complication such as memory leak.

The desired outcome is the resolution of the above problems.

1. Long-running, user-invoked tasks such as `test`, `run`, and `console` do not block BSP-origin queries and tasks.
2. From the time user issues in `test` to the test execution should be comparable to the current in-process execution, which keeps the `ClassLoader` warm for `scala-library`, and all `Test` dependency under turbo mode.
3. Leaky library usages in tests do not negatively affect the sbt server.
4. The system is capable handling at least the same workload as the status quo, which can execute multiple tests in parallel.

At least for the goal of sbt 2.0, I mostly want to tackle things that looks similar to `run` like take a bag of JARs and run them. `tests` are just a special case of `run` that doesn't take stdin.

>  It sounds like we want the following:
> - We want to always run every task as a `bgTask` (similar to Bazel's ephemeral workers).

No, most tasks should continue to run in the same JVM process as the sbt server for now. However, all tests should run like `bgRun` so one persistent worker can handle more than one test at a time.

> - Persistent workers are like `bgTask` runners, but they stick around after execution.

This behavior is configurable via `--persistent_worker` flag, and we could maybe recycle the worker process after some time.

> - Persistent workers should be an option, since it is a performance boost but also affects stability

Yea, so if there's a setting called `workerCommandLimit` or something, user can set that to `1` and you don't get the _persistence_ effect. Once we have that, I'd actually propose that we get rid of current in-process test runner and forked runner.

**Comment by eed3si9n** [comment]
If I may take a position a bit counter to the proposal (as Eugene 2), one thing we can observe is that there are two classes of tasks that are bundled up in one proposal.

- Async fire-and-forget tasks: `test` is a good example. We just want to run a bunch of tests in whatever order. Maybe we also want to see the log sometimes (Bazel doesn't show test logs unless you put a flag). Importantly, does NOT require stdin emulation!
- Synchronous, sequential task: `console` and `run` require human input, and thus could block for hours. The user could start `console` and go on a vacation. This also requires stdin emulation, which is possible (sbtn does it already) but messy. Imagine all the up arrows someone could type into Scala REPL. Those up arrows translate to some ANSI code, and gets sent to a virtual terminal, which then becomes a history lookup.

So my feedback to RFC-4 is to forget about stdin for persistence worker, and try doing that from the sbt shell and sbtn. Like you get a _description_ of what to run, like JARs in temp directory, and either program could run it using ForkMain. Very similar to `bgRun`, like @adpi2 has been saying.

Initially we should limit the scope to async fire-and-forget tasks scenario, since `test` blocking BSP is more pressing issue that would like to resolve.
